tests/cases/cancel.rs

```diff
@@ -0,0 +1,225 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::{Arc, Mutex};
+
+use futures::{future, stream as streams, Async, Future, Poll, Sink, Stream};
+use grpcio::*;
+use grpcio_proto::example::route_guide::*;
+use grpcio_proto::example::route_guide_grpc::*;
+
+type Handler<T> = Arc<Mutex<Option<Box<T>>>>;
+type BoxFuture = Box<Future<Item = (), Error = ()> + Send + 'static>;
+type RecordRouteHandler =
+    Handler<Fn(RequestStream<Point>, ClientStreamingSink<RouteSummary>) -> BoxFuture + Send>;
+type RouteChatHandler =
+    Handler<Fn(RequestStream<RouteNote>, DuplexSink<RouteNote>) -> BoxFuture + Send>;
+
+#[derive(Clone)]
+struct CancelService {
+    record_route_handler: RecordRouteHandler,
+    route_chat_handler: RouteChatHandler,
+}
+
+impl CancelService {
+    fn new() -> CancelService {
+        CancelService {
+            record_route_handler: Arc::new(Mutex::new(None)),
+            route_chat_handler: Arc::new(Mutex::new(None)),
+        }
+    }
+}
+
+impl RouteGuide for CancelService {
+    fn get_feature(&self, _: RpcContext, _: Point, sink: UnarySink<Feature>) {
+        // Drop the sink, client should receive Cancelled.
+        drop(sink);
+    }
+
+    fn list_features(&self, _: RpcContext, _: Rectangle, sink: ServerStreamingSink<Feature>) {
+        // Drop the sink, client should receive Cancelled.
+        drop(sink);
+    }
+
+    fn record_route(
+        &self,
+        ctx: RpcContext,
+        stream: RequestStream<Point>,
+        sink: ClientStreamingSink<RouteSummary>,
+    ) {
+        let handler = self.record_route_handler.lock().unwrap();
+        if handler.is_some() {
+            let f = (handler.as_ref().unwrap())(stream, sink);
+            ctx.spawn(f);
+        }
+    }
+
+    fn route_chat(
+        &self,
+        ctx: RpcContext,
+        stream: RequestStream<RouteNote>,
+        sink: DuplexSink<RouteNote>,
+    ) {
+        let handler = self.route_chat_handler.lock().unwrap();
+        if handler.is_some() {
+            let f = (handler.as_ref().unwrap())(stream, sink);
+            ctx.spawn(f);
+        }
+    }
+}
+
+fn check_cancel<S, T>(rx: S)
+where
+    S: Stream<Item = T, Error = Error>,
+{
+    match rx.into_future().wait() {
+        Err((Error::RpcFailure(s), _)) | Err((Error::RpcFinished(Some(s)), _)) => {
+            assert_eq!(s.status, RpcStatusCode::Cancelled)
+        }
+        Err((e, _)) => panic!("expected cancel, but got: {:?}", e),
+        Ok(_) => panic!("expected error, but got: Ok(_)"),
+    }
+}
+
+fn run_suite() -> (CancelService, RouteGuideClient, Server) {
```

Comment:
This name is a bit misleading. `run_suite` reads like it runs some given test suite, but instead it builds some common structures.